### PR TITLE
Signum with grad compression

### PR DESCRIPTION
**Pull request by yuxiangw**
## Description ##
Added a new grad compression scheme.
It keeps track of the momentum on every worker and sends the sign vector of the momentum.
This is a somewhat hacky version of doing distributed version of the Signum optimizer: https://github.com/apache/incubator-mxnet/pull/9220

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by eric-haibin-lin** [comment]
@rahul003

**Comment by rahul003** [comment]
Thanks for contributing to this. The implementation looks fine. Performance should be similar to 2bit compression, right?
But have you run convergence tests? How does it do on that front?

**Comment by yuxiangw** [comment]
Thanks @rahul003 for looking into this. This is used for a specific purpose --- running Signum over multiple GPU (and multiple machines) and illustrating the communication savings of this algorithm.  The convergence is guaranteed because it is not an approximation to other algorithms, it is the exact algorithm that we analyzed theoretically.

**Comment by yuxiangw** [comment]
@rahul003  The performance can be improved by a little bit I think (but not from the current implementation). I am thinking of changing the encoding and decoding to 1 bit per coordinate instead of 2 bits.  Are you able to help with that?

**Comment by rahul003** [comment]
How many states do you want the gradient to be stored in? (+,- and 0), or only (+ and -)? Currently your code uses three states, in which case we can't do 1bit.

(Also, one thing to keep in mind is that currently for distributed setting, gradients are compressed when sending from worker to server, but not from server to worker. A worker pulls the latest full weights that a server has. If it does for both, there would be two levels of compression and information loss, which I thought might hurt accuracy. I've not tried doing that yet. )

**Comment by yuxiangw** [comment]
Just + and -.  
RE:  gradients are compressed when sending from worker to server, but not from server to worker.

This is not ideal.  For K workers, you will only need log K bits per coordinate when broadcasting the weights  (come and ask why in person if this is not straightforward. I'm sitting near Mu).  Because you are sending only the differences to the workers. I thought Chao did that before passing over the project to you. Is it not the case?

Again, you will not lose accuracy when updating worker weights if you send the log K bits diff back.  Other things, such as the weight decay, should be done on each worker separately instead of communicating floating point numbers.

**Comment by rahul003** [comment]
It's fairly easy to change the compression to use 1bit. You need to change the bitmasks in the struct quantize_signum and dequantize_signum, and also change the start and end indices to cover 32 values in one call. 
You could use masks like 0x80, 0x40, 0x20, 0x10, 0x08, 0x04, 0x02, 0x01 for the values which are greater than 0. As an optimization don't encode negative values, and remove the second if statement in quantize. Let them be 0.

Yes you'll not lose accuracy if the server forwards the differences to worker. But the way mxnet kvstore currently works is that it pushes the full model to the workers. As against this, if we tried to send differences then we would need to send (num_workers * compressed_size) data as updates to each worker. This is because we can't merge the updates from different workers without decompressing, merging and then compressing them back. Also if we do this, then if num_workers becomes higher than the compression factor, then gradient compression would give no speedup.

**Comment by chinakook** [comment]
Microsoft made its [1-bit SGD](https://github.com/Microsoft/CNTK/tree/master/Source/1BitSGD) to open. I think someone can borrow some idea from that.

**Comment by shuo-ouyang** [comment]
> It's fairly easy to change the compression to use 1bit. You need to change the bitmasks in the struct quantize_signum and dequantize_signum, and also change the start and end indices to cover 32 values in one call.
> You could use masks like 0x80, 0x40, 0x20, 0x10, 0x08, 0x04, 0x02, 0x01 for the values which are greater than 0. As an optimization don't encode negative values, and remove the second if statement in quantize. Let them be 0. During dequantization change them to -1.
> 
> Yes you'll not lose accuracy if the server forwards the differences to worker. But the way mxnet kvstore currently works is that it pushes the full model to the workers. As against this, if we tried to send differences then we would need to send (num_workers * compressed_size) data as updates to each worker. This is because we can't merge the updates from different workers without decompressing, merging and then compressing them back. Also if we do this, then if num_workers becomes higher than the compression factor, then gradient compression would give no speedup.

@rahul003 

Hi, I have implemented a similar 1bit gradient compression algorithm with your advice. However, when I train resnet110 on CIFAR-10 dataset to compare my implementation to 2bit compression and no compression algorithms, I found that the speedup of gradient quantization for ResNet training seems not significantly. The training command and logs are shown as follows, and I deployed training tasks across four nodes (each node is equipped four K80 GPUs), in which one parameter server and three workers. Is there any incorrect setup in my training process?

ps: I use the example code in dictionary`example/image-classification`
mxnet version: 1.4.0
cuda: 8.0

**training command:**
```shell
python ../../tools/launch.py --launcher ssh -H hosts -s 1 -n 3 python train_cifar10.py --gc-type 2bit --gc-threshold 1 --kv-store dist_sync --num-epochs 200 --batch-size 128 --lr-step-epochs 100,150 --wd 0.0001 --lr 0.1 --lr-factor 0.1 --network resnet --gpus 0,1,2,3
```

**100th epoch:**
||  No Quantization   | 2bit Quantization  | 1bit Quantization  |
|:----:|  :----:  | :----:  |:----:  |
|time cost (second)| 19.27  | 19.777 |18.545 |
|validation accuracy| 0.89122  | 0.887921 |0.885871 |

**150th epoch:**
||  No Quantization   | 2bit Quantization  | 1bit Quantization  |
|:----:|  :----:  | :----:  |:----:  |
|time cost (second)| 18.73  | 22.357 |20.339 |
|validation accuracy| 0.92758  | 0.929688 |0.929109 |


**200th epoch:**
||  No Quantization   | 2bit Quantization  | 1bit Quantization  |
|:----:|  :----:  | :----:  |:----:  |
|time cost (second)| 19.048  | 18.846 |19.649 |
|validation accuracy| 0.929988  | 0.935397 |0.937500 |
